exoscale/__init__.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+import os.path
+import toml
+from .api.compute import ComputeAPI
+from .api.sos import SosAPI
+from .api.runstatus import RunstatusAPI
+
+_API_ENDPOINT_ENVVAR = "EXOSCALE_API_ENDPOINT"
+_API_KEY_ENVVAR = "EXOSCALE_API_KEY"
+_API_SECRET_ENVVAR = "EXOSCALE_API_SECRET"
+_SOS_REGION_ENVVAR = "EXOSCALE_SOS_REGION"
+
+
+class Exoscale(object):
+    """
+    An Exoscale API client.
+    """
+
+    def __init__(self):
+        self.api_endpoint = os.getenv(_API_ENDPOINT_ENVVAR, "")
+        self.api_key = os.getenv(_API_KEY_ENVVAR, "")
+        self.api_secret = os.getenv(_API_SECRET_ENVVAR, "")
+        self.sos_region = os.getenv(_SOS_REGION_ENVVAR, "")
+
+        self.compute = ComputeAPI(
+            endpoint=self.api_endpoint, key=self.api_key, secret=self.api_secret
+        )
+
+        if self.sos_region:
+            self.sos = SosAPI(
+                region=self.sos_region,
+                endpoint=self.api_endpoint,
```

Comment:
There are :)

* https://github.com/exoscale/python-exoscale/blob/dev/exoscale/api/sos.py#L15
* https://github.com/exoscale/python-exoscale/blob/dev/exoscale/api/runstatus.py#L17

Reply:
My understanding is that if you set `EXOSCALE_API_ENDPOINT`, it gets propagated as endpoint in all sub-services: compute, sos, runstatus.

Reply:
That's true, and this is something I should test to ensure this is not the case – but in practice the endpoint value is not supposed to be specified/overridden by users (this may happen in our internal context).

Reply:
Yep. For our internal override case it would still be convenient to override individually, e.g. by having `compute_api_endpoint`, `sos_api_endpoint` and `runstatus_api_endpoint` in the base API class.

Reply:
I could rewrite the `Exoscale` class' constructor this way:

```python
class Exoscale:
    """
    An Exoscale API client.
    """

    def __init__(self):
        self.api_key = os.getenv(_API_KEY_ENVVAR, "")
        self.api_secret = os.getenv(_API_SECRET_ENVVAR, "")

        self.compute = ComputeAPI(
            endpoint=os.getenv(_COMPUTE_API_ENDPOINT_ENVVAR, None),
            key=self.api_key,
            secret=self.api_secret
        )

        if self.sos_region:
            self.sos = SosAPI(
                region=os.getenv(_SOS_REGION_ENVVAR, None),
                endpoint=os.getenv(_SOS_API_ENDPOINT_ENVVAR, None),
                key=self.api_key,
                secret=self.api_secret,
            )

        self.runstatus = RunstatusAPI(
            endpoint=os.getenv(_RUNSTATUS_API_ENDPOINT_ENVVAR, None),
            key=self.api_key,
            secret=self.api_secret
        )
```

And then in each API submodule's constructor, set a default endpoint if `endpoint=None` is passed?

Reply:
Yep, sounds good!

Reply:
Fixed in bf10a01.